src/ray/object_manager/pull_manager.cc

```diff
@@ -350,24 +355,44 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
   // sending RPCs.
   if (!request.spilled_url.empty()) {
     const auto spilled_node_id = request.spilled_node_id;
+    // Set a flag on the request so we don't send duplicate restoration requests.
+    request.is_being_restored = true;
     restore_spilled_object_(
         object_id, request.spilled_url, spilled_node_id,
         [this, object_id, spilled_node_id](const ray::Status &status) {
+          auto it = object_pull_requests_.find(object_id);
+          if (it != object_pull_requests_.end()) {
+            // The pull request may no longer exist, e.g. it could have been cancelled.
+            it->second.is_being_restored = false;
```

Comment:
Is it guaranteed that after all callbacks are invoked after restoration, the pull requests are cancelled?

Reply:
Otherwise it is possible `is_being_restored` is still true after all of callbacks are completed right?

Reply:
>Is it guaranteed that after all callbacks are invoked after restoration, the pull requests are cancelled?

The pull request will be cancelled by higher-level code around task execution or fulfilling `ray.get()` or `ray.wait()`, but we don't want the pull requests to be cancelled right after restoration since we still need to pull the object.

>Otherwise it is possible is_being_restored is still true after all of callbacks are completed right?

There should be at most one `restore_spilled_object_` call, and therefore at most one callback, per pull request (due to our deduping on `request.is_being_restored`); after this callback executes, `request.is_being_restored` will be `false`, so the next time `TryToMakeObjectLocal` executes, it will pull the object.So I don't think that there's any way for `request.is_being_restored` to be true after the callback has completed.

Reply:
That did just make me think of something though... Upon a successful restoration, should we manually add the `spilled_node_id` to the locations for this request to be sure that the next `TryToMakeObjectLocal` won't trigger an unnecessary restoration request due to the object location subscription having not yet updated our view of the object locations yet?

Reply:
@rkooo567 I implemented what I think is more ideal behavior here, let me know if I'm missing something (if so I can revert that commit): https://github.com/ray-project/ray/pull/14660/commits/bf4a0df4fca53c3a3ab1bfbaac9a603318793792

Reply:
Sorry for being late! The current fix without this change is matching the previous behavior right (so this issue also happened before too)? Actually, can we just revert this and leave as TODO? I'd like to avoid adding more logic until we do refactoring.

Reply:
No worries! And the previous behavior was to infinitely retry restoration requests until the object location subscription returned the restored node location; this PR is supposed to prevent unnecessary restoration retries, so I think that this fix is required to meet that criteria.

Reply:
Can we just do this later? I really want to avoid adding more logic right now as this path should work very reliably in the short term (for testing). More logic is more prone to error. Let's add this logic once things are easier to understand after refactoring.

Reply:
I think the current fix is already an improvement that it will avoid unnecessary RPCs a lot.

Reply:
I don't really get that reasoning, this new logic is fixing what I believe to be an error in the long-polling protocol, which is definitely within the scope of this PR. When the restoration request successfully completes, we know that the object is in the object store on the spilled node but the object location subscription may be lagging behind the restoration RPC, so we try to add it to the object locations. This is equivalent to an object location change, so we try to make the object local.

If I'm mistaken in this being an error in the long-polling protocol, then lmk and I'll revert it, but I don't think that we should be resistant to new logic if it fixes an error.

Reply:
Discussed offline. @clarkzinzow mentioned this was necessary due to the new changes, so we will just keep it here.